### PR TITLE
[#18848][fix] Preserve self-configured CPU affinity on repeated calls

### LINKED ISSUE #18848: [Bug] configure_cpu_affinity() is not idempotent: a second call un-pins every thread
### Summary

`configure_cpu_affinity()` is not idempotent. Calling it twice in the same process un-pins every thread it pinned on the first call, silently reverting the NUMA optimization.

### Cause

The function decides whether affinity was set *by someone else* using only the current mask:

```python
all_cpus = list(range(psutil.cpu_count()))
constrained_affinity = (cpu_affinity != all_cpus)
```

It cannot tell an externally imposed constraint from the one it applied itself. On a second call the process is "constrained" — by its own previous NUMA pin — so with `TLLM_NUMA_AWARE_WORKER_AFFINITY` unset it takes the *un-constrain* branch, logs "Removing CPU affinity constraints", and rebinds every thread back to the full CPU set. The NUMA branch is then skipped, because `constrained_affinity` is true.

### Reproduction

Measured on a GB300 (Grace, 144 CPUs) with both functions extracted verbatim from `tensorrt_llm/llmapi/utils.py`, `psutil`/`get_numa_aware_cpu_affinity` shimmed onto real syscalls, six worker threads held alive across the whole run, `TLLM_NUMA_AWARE_WORKER_AFFINITY` unset:

```
start (unconstrained)   main=ALL   non-main: 0 on NUMA subset, 6 on ALL (of 6)
after 1st configure     main=NUMA  non-main: 6 on NUMA subset, 0 on ALL (of 6)
after 2nd configure     main=ALL   non-main: 0 on NUMA subset, 6 on ALL (of 6)
```

Emitted logs, in order:

```
INFO  Worker process ... CPU affinity set to [...] for optimal NUMA-aware scheduling (7/7 threads).
WARN  Worker process ... is affined to run on the following CPUs: [...]
WARN  Worker process ... has constrained CPU affinity but `TLLM_NUMA_AWARE_WORKER_AFFINITY` is not set. Removing CPU affinity constraints.
```

The second warning pair is the give-away: it reports the process as externally constrained when the constraint is self-inflicted.

### How it is reached

Any path that runs `configure_cpu_affinity()` twice in one process — most directly two sequential `LLM()` objects on the in-process worker path (`gather_generation_logits=True` or `TLLM_WORKER_USE_SINGLE_PROCESS=1`), a shape several in-tree tests already use.

### Severity

The end state matches the pre-#18701 baseline, so this is not a regression — but it silently reduces the feature to a no-op for the whole process, and the misleading "constrained CPU affinity" warning sends anyone debugging it in the wrong direction. Threads created between the two calls are a genuine small regression: they inherited the NUMA mask and are then widened, whereas previously they would have stayed pinned.

### Suggested direction

Track whether this process already applied a NUMA-aware mask (or compare the current mask against the NUMA-aware set) instead of treating any non-full mask as external. Whatever the fix, `constrained_affinity` should stop conflating "the user pinned us" with "we pinned ourselves".

### Context

Found while reviewing #18701; not introduced by it.

### PR DESCRIPTION
**Pull request by sylvesterkaczmarek**
## Description

Closes #18848.

Remember the PID and observed mask after applying NUMA affinity. Repeated calls can then reapply that mask to existing threads instead of mistaking it for an external constraint and unpinning them. A PID change or externally changed mask invalidates ownership. Explicit opt-in/opt-out behavior remains unchanged.

A partial bind records ownership only when the main thread's read-back mask is consistent with the requested CPUs. The function's policy documentation is updated; no dependencies or public API signatures change.

## Test Coverage

- Added policy regressions for repeated calls, device retargeting, external masks, opt-out, PID changes, failed binding, and retries.
- Added a Linux subprocess test covering threads created before and after the first pin.
- Local Linux validation: 13 tests passed. The harness executed the exact affinity functions from `utils.py` while excluding CUDA-dependent package initialization. NUMA discovery was stubbed; the subprocess exercised real Linux affinity system calls. This was not a full TensorRT-LLM package or GPU test run.
- Five regression cases failed on the original source.
- Repository pre-commit checks and `git diff --check` passed.

## PR Checklist

- [x] Reviewed the applicable contribution, coding, test, documentation, and dependency requirements.

## Dev Engineer Review

- `configure_cpu_affinity()` tracks the owning PID and applied CPU mask.
- Repeated calls reapply internally configured NUMA affinity instead of removing it.
- PID changes and external mask changes invalidate ownership.
- No public API, dependency, or default-policy changes are introduced.

## QA Engineer Review

- `tests/unittest/llmapi/test_cpu_affinity.py` covers repeated configuration, retargeting, external changes, opt-out behavior, PID changes, binding failures, retries, and real worker-thread mask preservation.
- Coverage addresses the reported idempotency regression and key ownership paths.
- No applicable `test-db/` or manual-QA list entry was identified.
- Coverage verdict: **sufficient**.

## Per-File QA Perspective

- `tensorrt_llm/llmapi/utils.py`: Verify repeated calls preserve NUMA affinity, newly created threads inherit the mask, external changes follow existing policy, and PID changes clear ownership. No API changes are expected.
- `tests/unittest/llmapi/test_cpu_affinity.py`: Unit and subprocess tests cover normal, retry, failure, external-change, opt-out, and thread-inheritance behavior. This unit test file does not require integration test-list registration.

### CODE REVIEW

File: tensorrt_llm/llmapi/utils.py [hunk 1]
```diff
@@ -761,17 +771,23 @@ def configure_cpu_affinity(device_id: int) -> None:
...
                 f"Worker process {pid} could not set the NUMA-aware CPU "
                 f"affinity of any thread. It will run without NUMA pinning, "
                 f"which may impact performance.")
         else:
+            effective_affinity = process.cpu_affinity()
+            # Other threads may bind successfully even if the main thread did
+            # not. Record only a read-back mask consistent with our request.
+            if effective_affinity and set(effective_affinity).issubset(
+                    numa_cpus):
+                _configured_cpu_affinity = (pid, frozenset(effective_affinity))
```
[the remark this answers is not in the document]
Comment (sylvesterkaczmarek) [reply tensorrt_llm/llmapi/utils.py:787]: Addressed in 824fb58. `test_proper_subset_readback_remains_owned_for_retry` requests [0, 1], reads back [0], calls configuration again, and verifies that the NUMA mask is reapplied without external-constraint warnings.
